File: mobile/ios/Sources/QuicTunnelMobileSdk/QuicTunnelMobileGrantPairingController.swift
import Foundation
import Security

public enum QuicTunnelMobileGrantPairingError: Error, Equatable {
    case invalidNonceByteCount(Int)
    case nonceGenerationFailed(OSStatus)
}

public final class QuicTunnelMobileGrantPairingController {
    public var options: FfiMobileGrantPairingOptions

    public static func makePairingOptions(
        controlRequestTimeoutMs: UInt64? = 5_000,
        controlMaxRetries: UInt32 = 2,
        controlRetryBackoffMs: UInt64 = 100
    ) -> FfiMobileGrantPairingOptions {
        var options = mobileGrantPairingOptionsWithDefaults()
        options.controlRequestTimeoutMs = controlRequestTimeoutMs
        options.controlMaxRetries = controlMaxRetries
        options.controlRetryBackoffMs = controlRetryBackoffMs
        return options
    }

    public static func generateNonce(byteCount: Int = 16) throws -> String {
        guard byteCount > 0 else {
            throw QuicTunnelMobileGrantPairingError.invalidNonceByteCount(byteCount)
        }
        var bytes = [UInt8](repeating: 0, count: byteCount)
        let status = bytes.withUnsafeMutableBytes { buffer -> OSStatus in
            guard let baseAddress = buffer.baseAddress else {
                return errSecParam
            }
            return SecRandomCopyBytes(kSecRandomDefault, buffer.count, baseAddress)
        }
        guard status == errSecSuccess else {
            throw QuicTunnelMobileGrantPairingError.nonceGenerationFailed(status)
        }

        return Data(bytes)
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .trimmingCharacters(in: CharacterSet(charactersIn: "="))
    }

    public init(options: FfiMobileGrantPairingOptions = mobileGrantPairingOptionsWithDefaults()) {
        self.options = options
    }

    public func start(
        invite: FfiMobileInvitePayload,
        clientId: String,
        requestedServices: [String],
        nonce: String? = nil,
        options explicitOptions: FfiMobileGrantPairingOptions? = nil
    ) throws -> FfiMobileGrantPairingSession {
        let resolvedNonce = nonce ?? (try Self.generateNonce())
        return try startMobileGrantPairing(
            invite: invite,
            clientId: clientId,
            requestedServices: requestedServices,
            nonce: resolvedNonce,
            options: explicitOptions ?? options
        )
    }

    public func pollOnce(
        pairing: FfiMobileGrantPairingSession,
        options explicitOptions: FfiMobileGrantPairingOptions? = nil
    ) throws -> FfiMobileGrantPairingPollResult {
        try pollMobileGrantPairingOnce(
            pairing: pairing,
            options: explicitOptions ?? options
        )
    }
}

File: mobile/android/src/main/java/dev/quictunnel/mobile/QuicTunnelMobileGrantPairingController.kt
package dev.quictunnel.mobile

import android.util.Base64
import java.security.SecureRandom
import uniffi.quic_tunnel_mobile_core.FfiMobileException
import uniffi.quic_tunnel_mobile_core.FfiMobileGrantPairingOptions
import uniffi.quic_tunnel_mobile_core.FfiMobileGrantPairingPollResult
import uniffi.quic_tunnel_mobile_core.FfiMobileGrantPairingSession
import uniffi.quic_tunnel_mobile_core.FfiMobileInvitePayload
import uniffi.quic_tunnel_mobile_core.mobileGrantPairingOptionsWithDefaults
import uniffi.quic_tunnel_mobile_core.pollMobileGrantPairingOnce
import uniffi.quic_tunnel_mobile_core.startMobileGrantPairing

class QuicTunnelMobileGrantPairingController(
    private val options: FfiMobileGrantPairingOptions = mobileGrantPairingOptionsWithDefaults(),
) {
    companion object {
        private const val DEFAULT_NONCE_BYTES = 16
        private val secureRandom = SecureRandom()

        @JvmStatic
        fun makePairingOptions(
            controlRequestTimeoutMs: ULong? = 5_000u,
            controlMaxRetries: UInt = 2u,
            controlRetryBackoffMs: ULong = 100u,
        ): FfiMobileGrantPairingOptions =
            mobileGrantPairingOptionsWithDefaults().apply {
                this.controlRequestTimeoutMs = controlRequestTimeoutMs
                this.controlMaxRetries = controlMaxRetries
                this.controlRetryBackoffMs = controlRetryBackoffMs
            }

        @JvmStatic
        fun generateNonce(byteCount: Int = DEFAULT_NONCE_BYTES): String {
            require(byteCount > 0) { "byteCount must be greater than zero" }
            val bytes = ByteArray(byteCount)
            secureRandom.nextBytes(bytes)
            return Base64.encodeToString(
                bytes,
                Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP,
            )
        }
    }

    @Throws(FfiMobileException::class)
    fun start(
        invite: FfiMobileInvitePayload,
        clientId: String,
        requestedServices: List<String>,
        nonce: String = generateNonce(),
        options: FfiMobileGrantPairingOptions = this.options,
    ): FfiMobileGrantPairingSession =
        startMobileGrantPairing(
            invite = invite,
            clientId = clientId,
            requestedServices = requestedServices,
            nonce = nonce,
            options = options,
        )

    @Throws(FfiMobileException::class)
    fun pollOnce(
        pairing: FfiMobileGrantPairingSession,
        options: FfiMobileGrantPairingOptions = this.options,
    ): FfiMobileGrantPairingPollResult =
        pollMobileGrantPairingOnce(
            pairing = pairing,
            options = options,
        )
}
